src/README.md

```diff
@@ -1206,6 +1228,51 @@ it, this can happen at any time after the garbage collector notices that
 it's no longer reachable and before the V8 isolate is torn down.
 See the [Oilpan documentation in Chromium][] for more details.
 
+When a cppgc-managed object is no longer reachable in the heap, its destructor
+will be invoked by the garbage collection, which can happen after the `Environment`
+is already gone, or after any object it references is gone. To ensure safety,
+the cleanups of a cppgc-managed object should adhere to different patterns,
+depending on what it needs to do:
+
+1. If it does not need to do any non-trivial, nor does its members, just use
+   the default destructor.
+2. If the cleanup relies on a living `Environment`, the class should use this
+   pattern in its class body:
+
+   ```cpp
+   ~MyWrap() { this->Clean(); }
+   void CleanEnvResource(Environment* env) override {
+     // Do cleanup that relies on a living Environemnt. This would be
+     // called by CppgcMixin::Clean() first during Environment shutdown,
+     // while the Environment is still alive. If the destructor calls
+     // Clean() again later during garbage collection that happens after
+     // Environment shutdown, CleanEnvResource() would be skipped, preventing
+     // invalid access to the Environment.
+   }
```

Comment:
I'm still in doubt if this example will entice people to access handles on the `node::Environment` given that it is passed in as a parameter, and ultimately call this `CleanEnvResource` through the destructor, which is invoked by cppgc.

I'm wondering if we should avoid creating a cppgc object that is short-lived than its associated env resources.

Reply:
Isn't the example in 3 enough to address this?  I can remove `Environment` as a parameter though I don't think it makes much of a difference - people can still easily reach out to `env()` if they want, just like in the case of BaseObject's destructors, even if there destructors are called as the first pass weak callback.

Reply:
Right, I think my comment was purely hypothetical.